src/model/engines/engine_base.nit

```diff
@@ -53,5 +76,79 @@ class Engine
 	fun compile(submission: Submission): Bool is abstract
 
 	# Run `test` for `submission`
-	fun run_test(submission: Submission, test: TestCase): TestResult is abstract
+	fun run_test(submission: Submission, test: TestCase): TestResult do
+		var res = new TestResult(test, submission)
+
+		var tdir = "test{submission.results.length + 1}"
+		# We get a subdirectory (a testspace) for each test case
+		var ws = submission.workspace.as(not null)
+		var ts = ws / tdir
+		ts.mkdir
+
+		# Prepare the input/output
+		var ifile = ts / "input.txt"
+		test.provided_input.write_to_file(ifile)
+		var ofile = ts / "output.txt"
+		var sfile = ts / "sav.txt"
+		test.expected_output.write_to_file(sfile)
+
+		var env = new TestEnvironment(ws, ifile, ofile, sfile, tdir)
+
+		execute_test(submission, res, env)
+
+		# Compare the result with diff
+		# TODO: some HTML-rich diff? Maybe client-side?
+		res.produced_output = ofile.to_path.read_all
+		var r = system("cd {ws} && diff -u {tdir}/sav.txt {tdir}/output.txt > {tdir}/diff.txt")
+		if r != 0 then
+			var out = (ts/"diff.txt").to_path.read_all
+			res.error = "Error: the result is not the expected one\n{out}"
+			return res
+		end
+
+		return res
+	end
+
+	# Executes a test on `submission`
+	fun execute_test(submission: Submission, res: TestResult, env: TestEnvironment) is abstract
+
+	# Make a workspace for the submission
+	fun make_workspace: nullable String do
+		# Get a workspace
+		# We need to create a unique working directory
+		# The following is not thread/process safe
+		var ws
+		var z = 0
+		loop
+			# Get a unique timestamp for this submission
+			var date = (new TimeT).to_i.to_s + "_" + z.to_s
+			ws = "out/{date}"
+			if not ws.file_exists then break
```

Comment:
Admittedly though, it is not super safe since two runs can make folders at the same time
But since Nitcorn is still mono-threaded, this should not be an issue for now